### ISSUE TITLE
Describe combination of Progressive calls and Shared registrations features

### ISSUE THREAD

**Issue by KSDaemon** [wamp-proto/wamp-proto#418]
We need to add more info about using this combination of features.

**Comment by oberstet** [comment]
This will be interesting! FWIW, I can already see 2 options with shared registrations and streaming calls:

the forwarding decision to which callee (of the multiple, due to shared registration) a call and its call-progress messages are directed to is:
* taken once, for the initial first CALL message received, and all subsequent CALL progress messages are forwarded to that same callee
* taken newly on each received CALL (progress) message for that call

**Comment by KSDaemon** [comment]
Btw, we already have some description of this here:

**Progressive calls and shared registration**

RPCs can have a multiple registrations (see `Shared Registration` feature) with different `<invocation_policies>`.
In this case `progressive` `CALLs` can be routed to different *Callees*, which can lead to unexpected results.
To bind `INVOCATIONs` to the same *Callee*, the *Caller* can specify the `sticky` option during `CALL`

`        CALL.Options.sticky|bool := true`


In this case *Dealer* must make a first `INVOCATION` based on `<invocation_policy>` and then route all 
subsequent `progressive` calls to the same *Callee*.

If binding all ongoing `progressive` calls to the same *Callee* is not required, *Caller* can set the `sticky` option to `FALSE`. 

If `CALL.Options.sticky` is not specified, it is treated like `TRUE`, so all `progressive`
calls go to the same *Callee*.

**Comment by KSDaemon** [comment]
I forgot about it :)))

**Comment by oberstet** [comment]
> I forgot about it :)))

=) yeah, `sticky` is pretty much exactly addressing the example I gave. I also like the default `true` .. makes sense, as for non-sticky, the app is better prepared to deal with this more complex situation.

anyways, I still think this issue is worth keep open for now .. we will likely need to add more details rgd specifically combining these 2 features.

a full duplex streaming call with shared registrations .. huh ... ;)

**Comment by oberstet** [comment]
> What I mean is: assume there are 10 Callees with a shared registration of procedure P, and one of the Callees returns an ERROR while the progressive call is still ongoing, the Router MUST interrupt all other 9 Callees (by sending them an INTERRUPT), and immediately send back ERROR to the original Caller

This is one option. "bubble up any error" and error the whole call

Another option: "swallow the error" and continue the call (without the failed Callee being invoked again). "swallow" could also mean, send an application error as a result to the Caller - what I would hesitate to do is forward the ERROR returned from one callee back to the caller, as ERROR should keep "concludes the whole call" semantics. always. IMO ..

**Comment by KSDaemon** [comment]
Well, maybe it is also a good point to become an attribute during the call. It's like `ALL|ANY` options kind of.

* Fail overall call on first error
* Continue till the end

And this only is applicable when the sticky is FALSE.

**Comment by oberstet** [comment]
yeah, agreed with using an attribute (so it can be controlled by the caller).

also "first" mode: yeah, that would correspond to "bubble up any error" and fail immediately.

not sure about "last": why would I want to fail the whole call only when the last error occurred? and how does a router know an error from an invocation is the last?

for the other option besides "first", what about "consume"?

as in "consume errors" and continue the call

Twisted has sth roughly similar with

* fireOnOneErrback
* consumeErrors

https://docs.twistedmatrix.com/en/stable/api/twisted.internet.defer.DeferredList.html

**Comment by KSDaemon** [comment]
What about this?
```
error_handling: 'fail_call'|'continue'
```

**Comment by oberstet** [comment]
one more bike shedding name detail: since this is only for error handling with `progress == true`, we could:

```
progress_error: 'fail_call'|'continue'
```

**Comment by Jopie64** [comment]
About the stickyness:
I think this should be determined by the callee, not by the caller. (So during RPC registration.)

Shared registration is also determined by the callee. And stickyness applies only to shared registration.
Also the complex non-sticky logic should be implemented on the callee side. It should be 'transparent' (😜) to the caller. It should not need to know where its progressive calls end up.

**Comment by oberstet** [comment]
> I think this should be determined by the callee, not by the caller. (So during RPC registration.)

actually, that's right, agreed. makes sense: the error handling policy applied is part of the service contract offered by the service.
